### ISSUE TITLE
create_task() returns task of a task

### ISSUE THREAD

**Issue by Mithras** [AppDaemon/appdaemon#921]
```
async def test(self):
    await self.sleep(1)
    return "ok"

task = self.create_task(self.test())
self.log(await task) # pending Task object
self.log(await (await task)) # "ok"
```

**Comment by Mithras** [comment]
I also have a few general questions:
1. Are there any naming conventions in Python for async methods? Or the only way to tell if something is async is to look at the source code?
2. Is there a specific reason to make methods with no awaits async? There are a lot of them in AppDaemon.
3. Why async methods in AppDaemon are executed even when I don't await them? I kinda expect them to do nothing until awaited and/or passed to create_task().

**Comment by ReneTode** [comment]
i dont have answers about 1 and 3, but about 2:

in general, if you dont have a specific reason to use async, then dont use it.
appdeamon is designed in sync, and async is recently added to accomodate situations where async is absolutely needed.

as the docs state: async is high advanced AD use and it should only be used if you really understand async, know what you do and understand that using async brings more risks that apps influence each other.

**Comment by Mithras** [comment]
Async does require a bit more digging into but I don't necessarily agree with "async is high advanced AD use". I've just finished migrating everything I have to AppDaemon 4 and async and my apps were never that simple and clean looking. And everything seems to be significantly faster on top of that. I can't tell if it's faster because of async or just because AppDaemon 4 is so much faster though.
I really hope some day (AppDaemon 5?) async will be the only option. Old model doesn't give you any benefits. I can still do run_in_executor() if I need to do some heavy lifting.

**Comment by ReneTode** [comment]
AD 4 is faster.

and there are advantages from using sync over async.
sync apps runn all extremewly separated, and sperated from the AD core.
errors in async apps can stop/break/ hold up the entire AD core, so all apps.
errors in sync apps in general only stop/break/hod up that single app and all other apps will run as usual.

i cant say if AD will run in async only at some point, but i dont think it ever will. or at least it shouldnt.

**Comment by Mithras** [comment]
> errors in async apps can stop/break/ hold up the entire AD core, so all apps.

I'm yet to see that happen in AD 4. And believe me I had a lot of errors while rewriting my 30+ apps.

> errors in sync apps in general only stop/break/hod up that single app and all other apps will run as usual.

The same as async apps. The only way to hold anything is to block the thread. Exception don't block anything.

> i cant say if AD will run in async only at some point, but i dont think it ever will. or at least it shouldnt.

How about making everything async and allow people to opt-in for a separate pinned thread for an app and run asyncio loop in it? This way people who want to be "extremewly separated" for whatever reason, can do that while still having all the benefits of async.

**Comment by ReneTode** [comment]
> How about making everything async and allow people to opt-in for a separate pinned thread for an app which will be used to run asyncio loop in? This way people who want to be "extremely separated" for whatever reason, can do that while still having all the benefits of async.

i am not a dev anymore, so i cant say anything about it.
but making AD async by default would break all sync apps.
and it would make everything way more complicated for new users.

> The only way to hold anything is to block the thread.
thats about it. it depends on what happens. mostly things that are overseen, unexpected.
blocking a sync app in that case would only block an app, and not the core.

i personally dont see any benifits from async apps. ;) 
but like i daid, i am no dev anymore.

**Comment by dlashua** [comment]
@Mithras 

There are SOME naming conventions in Python for async. Generally speaking, if there are two methods that do the same thing, one with async, and one without, the async version is prefixed with async_. However, for applications that are primarily async (as AppDaemon is, in its core) this would mean prefixing EVERY method with "async_".

The methods available as "self.method" in an AppDaemon app are technically async methods. However, if used within AppDaemon and in a synchronous method, they will act like a synchronous method. This is by design to allow users to program their apps synchronously, which is often easier. AppDaemon is also Threaded, so, those executions are not actually happening synchronously, but the end result to the user appears that way.

You are seeing a task of a task because of the way you are using "create_task".

If you have this:
```
async def test(self):
  await self.sleep(1)
  return "ok"
```

and you are trying to call it from an async method, you do not need to "create_task". Simply:
```
async def something(self):
  await self.test()
```

if you are trying to call it from a sync method, then create_task will help:
```
def something(self):
  task = self.create_task(self.test())
  # you cannot use `await task`, because you are not in an async method

  # do some things

  if task.done():
    # do something because the task has completed
```

create_task() is meant to be used when you want to schedule background execution of a coroutine and when you will write code to check on it later (or simply run it with no need to check on it). This can be done with a run_in(), for instance, to check the status every few minutes/seconds until the task has completed and then do something with task.result().

**Comment by Mithras** [comment]
According to documentation self.create_task() is supposed to be an alternative to asyncio.create_task() so that AppDaemon can do all the cancelations and stuff. The problem is self.create_task() has diffenet interface and can not be used as an alternative to asyncio.create_task().
If you want to understand why it is an issue check out this https://github.com/Mithras/ha/blob/tmp/appdaemon/apps/globals.py#L26
In short, I need to start doing something async in __init__ which cannot be async. I'm using create_task() to start that something and then I want to await the task but I can't because it a task of a task. I can of course await await the task but don't think it's the right thing to do. I don't think it's ever valid to return task of a task anyway.

**Comment by dlashua** [comment]
I see what you're saying. You're using this correctly (creating a task in a sync method (__init__) and wanting AD to manage clean up of it. It should be returning a useful Task. However, our method decorator may be obscuring that.

I see that, in your class, you've provided your own create_task() method. Does this method work as you need it to? Also, which version of Python are you using? As we support Python 3.6, we are unable to use asyncio.create_task() and, instead, use asyncio.ensure_future().

I will do some testing on this shortly and get back to you. Thanks for taking the time to explain the issue.

**Comment by Mithras** [comment]
> I see that, in your class, you've provided your own create_task() method. Does this method work as you need it to? Also, which version of Python are you using? As we support Python 3.6, we are unable to use asyncio.create_task() and, instead, use asyncio.ensure_future().

I just copy-pasted original AD create_task and removed `async `modifier. I replaced `ensure_future()` with `create_task()` for readability only. I'm pretty sure it doesn't matter here which one to use. I'm using latest version of hassio + AppDaemon 4 addon. 
```
pi@hassio:~ $ python3 --version
Python 3.7.3
```

> I will do some testing on this shortly and get back to you. Thanks for taking the time to explain the issue.
Thank you!
